src/graph/sampler/metapath.cc

```diff
@@ -39,11 +39,20 @@ RandomWalkTracesPtr MetapathRandomWalk(
   const dgl_type_t *etype_data = static_cast<dgl_type_t *>(etypes->data);
   const dgl_id_t *seed_data = static_cast<dgl_id_t *>(seeds->data);
 
+  // vertices[seed][concatenated_traces]
+  std::vector< std::vector<dgl_id_t> > vertices_per_seed(num_seeds);
+  // trace_lengths[seed][trace_counts[seed]]
+  std::vector< std::vector<size_t> > trace_lengths_per_seed(num_seeds);
+  // trace_counts[seed]
+  std::vector<size_t> trace_counts(num_seeds);
   std::vector<dgl_id_t> vertices;
-  std::vector<size_t> trace_lengths, trace_counts;
+  std::vector<size_t> trace_lengths;
 
   // TODO(quan): use omp to parallelize this loop
-  for (uint64_t seed_id = 0; seed_id < num_seeds; ++seed_id) {
+#pragma omp parallel for
```

Comment:
Be aware of `hg->SuccVec`. It will materialize CSR if it does not exist and cause data race.

Reply:
Is it doing so now?  I couldn't find any reference to lazy materialization of CSR for heterographs.